File: SDKs/javascript/frameworks/app_kit/menu_item.js
/* 
 * menu_item.js
 * vienna
 * 
 * Created by Adam Beynon.
 * Copyright 2009 Adam Beynon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

include('app_kit/view');

var NSMenuItem = NSObject.extend({
  
  _menu: null,
  _subMenu: null,
  _title: null,
  _keyEquivalent: null,
  _keyEquivalentModifierMask: null,
  _mnenomicLocation: null,
  _state: null,
  _onStateImage: null,
  _offStateImage: null,
  _target: null,
  _action: null,
  _tag: null,
  _extraData: null,
  
  _isEnabled: null,
  _isHidden: null,
  
  init: function() {
    this._super();
    this._title = "";
    return this;
  },
  
  initWithTitle: function(itemName, action, keyEquivalent) {
    this.init();
    this._title = itemName;
    this._action = anAction;
    this._keyEquivalent = keyEquivalent;
    
    this._isEnabled = true;
    this._isHidden = false;
    
    return this;
  },
  
  initWithCoder: function(aCoder) {
    this._title = aCoder.decodeObjectForKey("NSTitle");
    this._keyEquivalent = aCoder.decodeObjectForKey("NSKeyEquiv");
    this._keyEquivalentModifierMask = aCoder.decodeIntForKey("NSKeyEquivModMask");
    this._menu = aCoder.decodeObjectForKey("NSMenu");
    this._submenu = aCoder.decodeObjectForKey("NSSubmenu");
    this._isEnabled = true;
    this._isHidden = false;
    return this;
  },
  
  setMenu: function(aMenu) {
    this._menu = aMenu;
  },
  
  menu: function() {
    return this._menu;
  },
  
  hasSubmenu: function() {
    return this._submenu ? true : false;
  },
  
  setSubmenu: function(submenu) {
    this._submenu = submenu;
  },
  
  submenu: function() {
    return this._submenu;
  },
  
  setTitle: function(aString) {
    this._title = aString;
  },
  
  title: function() {
    return this._title;
  },
  
  setAttributedTitle: function(aString) {
    this._title = aString;
  },
  
  /**
    Returns an NSAttributedString ready for drawing for the title.
  */
  attributedTitle: function() {
    var attributes = NSDictionary.create();
		
		// font
		if (this._menu.menuView().font())
			attributes.setObjectForKey(this._menu.menuView().font(), NSFontAttributeName);
		
		// textColor
		if (this.isEnabled()) {
		  if (this.isHighlighted())
		    attributes.setObjectForKey(NSColor.selectedMenuItemTextColor(), NSForegroundColorAttributeName);
		  else
			  attributes.setObjectForKey(NSColor.colorWithCalibratedRGBA(0.916, 0.916, 0.916, 1.0), NSForegroundColorAttributeName);
		}
		else {
			attributes.setObjectForKey(NSColor.disabledControlTextColor(), NSForegroundColorAttributeName);
		}
		
		return NSAttributedString.create('initWithStringAndAttributes', this._title, attributes);
  },
  
  /**
    Added Vienna convenience method..
    
    Retuns an NSAttributedString ready for drawing key equiv
  */
  attributedKeyEquivalent: function() {
    var attributes = NSDictionary.create();
    
    // font
    if (this._menu.menuView().font())
      attributes.setObjectForKey(this._menu.menuView().font(), NSFontAttributeName);
    
    // textcolor
    if (this.isEnabled()) {
		  if (this.isHighlighted())
		    attributes.setObjectForKey(NSColor.selectedMenuItemTextColor(), NSForegroundColorAttributeName);
		  else
			  attributes.setObjectForKey(NSColor.colorWithCalibratedRGBA(0.8, 0.8, 0.8, 1.0), NSForegroundColorAttributeName);
		}
		else {
			attributes.setObjectForKey(NSColor.disabledControlTextColor(), NSForegroundColorAttributeName);
		}
		
    // mask
    var theKeyEquiv = "";
    var modMask = this.keyEquivalentModifierMask();
    if (modMask & NSShiftKeyMask)
		  theKeyEquiv = theKeyEquiv + "⇑";
		
		theKeyEquiv = theKeyEquiv + this._keyEquivalent.toUpperCase()
		
		return NSAttributedString.create('initWithStringAndAttributes', theKeyEquiv, attributes);
  },
  
  isSeparatorItem: function() {
    // this should return true if no title & no image?!?!?!?
    return this._title ? false : true;
  },
  
  setKeyEquivalent: function(keyEquivalent) {
    
  },
  
  keyEquivalent: function() {
    return this._keyEquivalent;
  },
  
  setKeyEquivalentModifierMask: function(mask) {
    
  },
  
  keyEquivalentModifierMask: function() {
    return this._keyEquivalentModifierMask;
  },
  
  userKeyEquivalent: function() {
    
  },
  
  setImage: function(menuImage) {
    
  },
  
  image: function() {
    
  },
  
  setState: function(state) {
    
  },
  
  state: function() {
    
  },
  
  setOnStateImage: function(image) {
    
  },
  
  onStateImage: function() {
    
  },
  
  setOffStateImage: function(image) {
    
  },
  
  offStateImage: function() {
    
  },
  
  setMixedStateImage: function(image) {
    
  },
  
  mixedStateImage: function() {
    
  },
  
  setEnabled: function(flag) {
    this._isEnabled = flag;
  },
  
  isEnabled: function() {
    return this._isEnabled;
  },
  
  setAlternate: function(isAlternate) {
    
  },
  
  isAlternate: function() {
    
  },
  
  setIndentationLevel: function(level) {
    
  },
  
  indentationLevel: function() {
    
  },
  
  setTarget: function(aTarget) {
    
  },
  
  target: function() {
    
  },
  
  setAction: function(anAction) {
    
  },
  
  action: function() {
    
  },
  
  setTag: function(anInt) {
    
  },
  
  tag: function() {
    
  },
  
  setRepresentedObject: function(anObject) {
    
  },
  
  representedObject: function() {
    
  },
  
  setView: function(aView) {
    
  },
  
  view: function() {
    
  },
  
  isHighlighted: function() {
    if (this._menu.highlightedItem() == this)
      return true;
    
    return false;
  },
  
  setHidden: function(hidden) {
    
  },
  
  isHidden: function() {
    
  },
  
  isHiddenOrHasHiddenAncestor: function() {
    
  },
  
  setToolTip: function(toolTip) {
    
  },
  
  toolTip: function() {
    
  }
});

File: lib/vienna/app_kit/lib/menu/menu_item.rb
# 
# menu_item.rb
# vienna
# 
# Created by Adam Beynon.
# Copyright 2009 Adam Beynon.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

module Vienna
  
  class MenuItem
    
    def MenuItem.separator_item
      # return new separator item
    end
    
    def initialize(title, action, key_equivalent)
      @title = title
      @action = action
      @key_equivalent = key_equivalent
    end
    
    def menu=(menu)
      @menu = menu
    end
    
    def menu
      @menu
    end
    
    def has_submenu?
      @submenu ? true : false
    end
    
    def submenu=(submenu)
      @submenu = submenu
    end
    
    def submenu
      @submenu
    end
    
    def parent_item
      # parent menuitem (not menu)
    end
    
    def title=(title)
      @title = title
    end
    
    def title
      @title
    end
    
    def attributed_title=(string)
      @attributed_title = string
    end
    
    def attributed_title
      @attributed_title
    end
    
    def separator_item?
      false
    end
    
    def key_equivalent=(key_equivalent)
      @key_equivalent = key_equivalent
    end
    
    def key_equivalent
      @key_equivalent
    end
    
    def key_equivalent_modifier_mask=(mask)
      @key_equivalent_modifier_mask = mask
    end
    
    def key_equivalent_modifier_mask
      @key_equivalent_modifier_mask
    end
    
    def image=(an_image)
      @image = an_image
    end
    
    def image
      @image
    end
    
    def state=(a_state)
      @state = a_state
    end
    
    def state
      @state
    end
    
    def on?
      @state == :on
    end
    
    def off?
      @state == :off
    end
    
    def mixed?
      @state == :mixed
    end
    
    def on_state_image=(image)
      @on_state_image = image
    end
    
    def on_state_image
      @on_state_image
    end

    def off_state_image=(image)
      @off_state_image = image
    end
    
    def off_state_image
      @off_state_image
    end

    def mixed_state_image=(image)
      @mixed_state_image = image
    end
    
    def mixed_state_image
      @mixed_state_image
    end
    
    def enabled=(flag)
      @enabled = flag
    end
    
    def enabled?
      @enabled
    end
    
    def alternate=(alternate)
      @alternate = alternate
    end
    
    def alternate?
      @alternate
    end
    
    def indentation_level=(indentation_level)
      @indentation_level = indentation_level
    end
    
    def indentation_level
      @indentation_level
    end
    
    def target=(a_target)
      @target = a_target
    end
    
    def target
      @target
    end
    
    def action=(an_action)
      @action = an_action
    end
    
    def action
      @action
    end
    
    def on_action(&block)
      obj = Object.new
      obj.instance_variable_set('@action', block)
      def obj.menu_item_action(sender)
        @action.call(sender)
      end
      self.action = :menu_item_action
      self.target = obj
    end
    
    def tag=(a_tag)
      @tag = a_tag
    end
    
    def tag
      @tag
    end
    
    def represented_object=(an_object)
      @represented_object = an_object
    end
    
    def represented_object
      @represented_object
    end
    
    def view=(a_view)
      @view = a_view
    end
    
    def view
      @view
    end
    
    
    def highlighted?
      false
    end
    
    def hidden=(hidden)
      @hidden = hidden
    end
    
    def hidden?
      @hidden
    end
    
    def hidden_or_has_hidden_ancestor?
      @hidden
    end
    
    def tooltip=(tooltip)
      @tooltip = tooltip
    end
    
    def tooltip
      @tooltip
    end
  end
  
  class View < Responder
    
    def enclosing_menu_item
      nil
    end    
  end
end
